src/doorman/Fd.cpp

```diff
@@ -0,0 +1,86 @@
+#include "Fd.hpp"
+
+#include <fmt/format.h>
+
+using namespace fmt::literals;
+
+void Fd::write(const void *data, size_t length) const {
+    if (!is_open())
+        throw std::runtime_error("Write called on invalid file descriptor");
+    auto num = ::write(fd_, data, length);
+    if (num < 0)
+        throw fmt::system_error(errno, "Unable to write to {}", fd_);
+    if (static_cast<size_t>(num) != length)
+        throw std::runtime_error("Truncated write to file descriptor {}"_format(fd_));
+}
+
+void Fd::writev(gsl::span<const iovec> io_vecs) const {
+    if (!is_open())
+        throw std::runtime_error("Write called on invalid file descriptor");
+    auto num = ::writev(fd_, io_vecs.data(), io_vecs.size());
+    if (num < 0)
+        throw fmt::system_error(errno, "Unable to write to {}", fd_);
+
+    size_t length = 0;
+    for (auto &v : io_vecs)
+        length += v.iov_len;
+
+    if (static_cast<size_t>(num) != length)
+        throw std::runtime_error("Truncated write to file descriptor {}"_format(fd_));
```

Comment:
This "..."_format(fd_) looks weird to me. I'd kind of expect a comma there. Is _format some macro that emits one (or more) things with a comma in front?
Related: why do you use _format(fd_) in some places, and '"... {}", fd_' in others? Should they be consistent?

Reply:
The `_format` suffix is a UDL (user defined literal). I _thought_ I was consistent in my use of formatting but there are two spots where I use the `fmt::format("...")` version (but not in this PR...) Do you have a link to where you  mean? I agree consistency (and will fix the ones in `Channel.cpp` that I found in the appropriate PR).

---

So! Behind the scenes, a UDL is a suffix you can put on a literal to turn it into any object. `fmtlib` (which is being adopted into C++20) makes a format object when you say `"blah"_format` -- basically it's the same as `createFormattingObject("blah")`. The object has an `operator()` which then formats a string with the given params. So:

```cpp
auto x = "Hello {}"_format(world);
```

is equivalent to:

```cpp
auto format_obj = MakeAFormatObj("Hello {}");
auto x = format_obj(world);
```

It looks weird BUT  the compiler can _compile time check_ the parameters: https://godbolt.org/z/PY9qMM